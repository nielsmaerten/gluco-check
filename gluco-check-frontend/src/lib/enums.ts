export enum BloodGlucoseUnit {
  mgdl = "mg/dl",
  mmol = "mmol/L",
}

export enum NightscoutBloodGlucoseUnit {
  mgdl = "mg/dl",
  mmol = "mmol",
}

// @TODO: once this repo is properly part of the workspace,
// pull this in from gluco-check-core instead
// These are mapped to translations directly by case name,
// additions or changes here need to be reflected there
export enum DiabetesMetric {
  Everything = "everything",
  BloodSugar = "blood sugar",
  InsulinOnBoard = "insulin on board",
  CarbsOnBoard = "carbs on board",
  SensorAge = "sensor age",
  CannulaAge = "cannula age",
  PumpBattery = "pump battery",
  PumpReservoir = "pump reservoir",
}

export enum AvailableLanguage {
  English = "en",
  Dutch = "nl",
}
