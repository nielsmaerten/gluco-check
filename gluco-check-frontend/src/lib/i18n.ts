/* istanbul ignore file */
import i18n from "i18next";
import Backend from "i18next-http-backend";
import { initReactI18next } from "react-i18next";
import LanguageDetector from "i18next-browser-languagedetector";

i18n
  .use(Backend)
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    fallbackLng: (code) => {
      const fallback = process.env.REACT_APP_I18N_FALLBACK_LANGUAGE || "en-US";
      return [fallback];
    },
    detection: {
      // this is just the defaults with navigator moved to the front of the line
      order: [
        "navigator",
        "querystring",
        "cookie",
        "localStorage",
        "sessionStorage",
        "htmlTag",
        "path",
        "subdomain",
      ],
    },
    debug: process.env.REACT_APP_I18N_DEBUG === "true" || false,
    interpolation: {
      escapeValue: false,
    },
  });

i18n.on("languageChanged", (lng) => {
  document.documentElement.setAttribute("lang", lng);
});

export default i18n;
