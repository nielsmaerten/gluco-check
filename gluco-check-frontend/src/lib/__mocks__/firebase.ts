import { BloodGlucoseUnit, DiabetesMetric } from "../enums";
import { GlucoCheckUserDocument } from "../types";

export const mockUser: Partial<firebase.User> = {
  email: "example@example.com",
  uid: "uid123",
};

export const mockUserDocument: GlucoCheckUserDocument = {
  heardDisclaimer: false,
  defaultMetrics: [DiabetesMetric.BloodSugar],
  glucoseUnit: BloodGlucoseUnit.mgdl,
  nightscout: {
    url: "https://example.com",
    token: "nstoken123",
  },
};
