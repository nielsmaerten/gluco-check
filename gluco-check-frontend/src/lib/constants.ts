import { BloodGlucoseUnit } from "./enums";

// firestore
export const FIRESTORE_FIELD_PATH_GLUCOSE_UNITS = "glucoseUnit";
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT = "nightscout";
export const FIRESTORE_FIELD_PATH_DEFAULT_METRICS = "defaultMetrics";
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT_URL = `${FIRESTORE_FIELD_PATH_NIGHTSCOUT}.url`;
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT_TOKEN = `${FIRESTORE_FIELD_PATH_NIGHTSCOUT}.token`;
export const FIRESTORE_FIELD_HEARD_DISCLAIMER = "heardDisclaimer";

// gluco check
export const NIGHTSCOUT_VALIDATION_ENDPOINT_URL =
  process.env.REACT_APP_NIGHTSCOUT_VALIDATION_ENDPOINT_URL || "";

// configuration
export const ALERT_AUTOHIDE_DURATION = 7000;
export const VALIDATION_DEBOUNCE_DURATION = 500;
export const DEFAULT_ONBOARDING_CAROUSEL_INTERVAL = 5000;
export const DEFAULT_GLUCOSE_UNITS = BloodGlucoseUnit.mmol;

// urls
export const TERMS_AND_CONDITIONS_URL = "/terms";
export const HEALTH_DISCLAIMER_URL = "/terms";
export const PRIVACY_URL = "/terms";
export const NIGHTSCOUT_PROJECT_URL = "http://nightscout.info/";
export const FAQS_URL = "/faq";
export const GLUCO_CHECK_GITHUB_URL =
  "https://github.com/nielsmaerten/gluco-check";

export const APP_DEBUG = process.env.REACT_APP_DEBUG === "true" || false;
