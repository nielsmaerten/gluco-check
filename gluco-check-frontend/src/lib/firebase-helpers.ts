import { BloodGlucoseUnit, DiabetesMetric } from "./enums";
import { GlucoCheckUserDocument } from "./types";

export const FIRESTORE_DEFAULT_SET_OPTIONS: firebase.firestore.SetOptions = {
  mergeFields: [
    "defaultMetrics",
    "glucoseUnit",
    "nightscout.token",
    "nightscout.url",
  ],
};

export const DEFAULT_USER_DOCUMENT: GlucoCheckUserDocument = {
  defaultMetrics: [DiabetesMetric.BloodSugar],
  glucoseUnit: BloodGlucoseUnit.mgdl,
  nightscout: {
    token: "",
    url: "",
  },
};

export const getDocumentPathForUser = (user: firebase.User): string => {
  return `users/${user.email}`;
};
