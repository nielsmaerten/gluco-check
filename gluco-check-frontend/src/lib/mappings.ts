import { BloodGlucoseUnit, NightscoutBloodGlucoseUnit } from "./enums";

export const NightscoutBloodGlucoseUnitMapping = {
  [BloodGlucoseUnit.mgdl]: NightscoutBloodGlucoseUnit.mgdl,
  [BloodGlucoseUnit.mmol]: NightscoutBloodGlucoseUnit.mmol,
};
