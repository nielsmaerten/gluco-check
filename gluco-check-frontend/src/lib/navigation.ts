import { AvailableLanguage } from "./enums";

const LANGUAGE_INDEX = 1;
const SLASH = "/";

export const replaceLanguageInPathname = (
  pathname: string,
  newLanguage: string,
  supportedLanguages: string[] = Object.values(AvailableLanguage)
) => {
  const splitRoute = pathname.split(SLASH);
  if (supportedLanguages.includes(newLanguage)) {
    splitRoute[LANGUAGE_INDEX] = newLanguage;
    return splitRoute.join(SLASH);
  }
  return pathname;
};
