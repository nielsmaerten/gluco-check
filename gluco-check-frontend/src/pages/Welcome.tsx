import React from "react";
import { useTranslation } from "react-i18next";
import {
  Button,
  Container,
  Grid,
  makeStyles,
  Typography,
} from "@material-ui/core";
import Onboarding from "../components/Onboarding";
import Boilerplate from "../components/Boilerplate";
import { auth } from "../lib/firebase";
import { useHistory } from "react-router-dom";

export const handleSignoutClicked = () => {
  auth.signOut();
};

const useStyles = makeStyles((theme) => ({
  container: {
    paddingTop: theme.spacing(2),
    textAlign: "center",
  },
  subtitle: {
    marginBottom: theme.spacing(2),
    textAlign: "center",
    [theme.breakpoints.down("sm")]: {
      marginBottom: 0,
      "& h2": {
        fontSize: "1.125rem",
        lineHeight: 1.6,
      },
    },
  },
  bodyContent: {
    "& p": {
      marginBottom: theme.spacing(2),
    },
    "& ul li": {
      listStyle: "none",
      marginBottom: theme.spacing(1),
      fontWeight: 600,
    },
  },
  contentGrid: {
    minHeight: "80vh",
  },
  boilerplate: {
    textAlign: "center",
    marginTop: theme.spacing(2),
    "& p": {
      marginBottom: theme.spacing(1),
      [theme.breakpoints.down("sm")]: {
        fontSize: ".7rem",
      },
    },
  },
  ctaContainer: {
    "&.MuiGrid-item": {
      padding: "0",
    },
  },
}));

function Welcome() {
  const classes = useStyles();
  const { t, i18n } = useTranslation();
  const history = useHistory();

  return (
    <Grid
      className={classes.contentGrid}
      container
      direction="column"
      justify="center"
      alignItems="center"
      spacing={2}
      wrap="nowrap"
    >
      <Grid item className={classes.subtitle}>
        <Container maxWidth="sm">
          <Typography variant="h5" component="h2">
            {t("welcome.subtitle")}
          </Typography>
        </Container>
      </Grid>
      <Grid item>
        <Onboarding />
      </Grid>
      <Grid item>
        <Button
          variant="contained"
          color="primary"
          data-testid="settings-button"
          onClick={() => {
            history.push(`/${i18n.language}/settings`);
          }}
        >
          {t("welcome.cta")}
        </Button>
      </Grid>
      <Grid item className={classes.boilerplate}>
        <Boilerplate handleSignoutClicked={handleSignoutClicked} />
      </Grid>
    </Grid>
  );
}

export default Welcome;
