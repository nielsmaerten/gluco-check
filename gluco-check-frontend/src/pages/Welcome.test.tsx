import React from "react";
import { cleanup, render } from "@testing-library/react";
import { axe } from "jest-axe";
import Welcome, { handleSignoutClicked } from "./Welcome";

import { useHistory, BrowserRouter } from "react-router-dom";
jest.mock("react-router-dom", () => {
  const originalModule = jest.requireActual("react-router-dom");
  return {
    ...originalModule,
    useHistory: jest.fn(),
  };
});
const mockUseHistory = useHistory as jest.Mock;

const mockLanguage = "en";
jest.mock("react-i18next", () => ({
  useTranslation: () => {
    return {
      t: jest.fn().mockImplementation((i) => {
        return i;
      }),
      i18n: {
        language: mockLanguage,
      },
    };
  },
  Trans: function (props: any) {
    return <span>{props.i18nKey}</span>;
  },
}));

import { auth } from "../lib/firebase";
import userEvent from "@testing-library/user-event";
jest.mock("../lib/firebase.ts", () => {
  return {
    auth: {
      signOut: jest.fn(),
    },
  };
});

jest.mock("../components/Onboarding", () => {
  return {
    __esModule: true,
    default: () => {
      return <>Onboarding</>;
    },
  };
});

jest.mock("../components/Boilerplate", () => {
  return {
    __esModule: true,
    default: () => {
      return <>Boilerplate</>;
    },
  };
});

afterEach(cleanup);

describe("Welcome page", () => {
  const mockHistoryPush = jest.fn();
  beforeEach(() => {
    mockUseHistory.mockReturnValue({
      push: mockHistoryPush,
    });
  });
  it("renders the component", async () => {
    expect.assertions(2);
    const { container, getByTestId } = render(
      <BrowserRouter>
        <Welcome />
      </BrowserRouter>
    );
    const settingsButton = getByTestId("settings-button");
    await userEvent.click(settingsButton);
    expect(mockHistoryPush).toBeCalled();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("has no axe violations", async () => {
    const { container } = render(
      <BrowserRouter>
        <Welcome />
      </BrowserRouter>
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("handleSignoutClicked", () => {
  beforeEach(() => {
    (auth.signOut as jest.Mock).mockReset();
  });
  it("Logs the user out when called", () => {
    handleSignoutClicked();
    expect(auth.signOut as jest.Mock).toHaveBeenCalled();
  });
});
