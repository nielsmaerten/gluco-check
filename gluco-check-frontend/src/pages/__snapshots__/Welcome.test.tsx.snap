// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome page renders the component 1`] = `
<div
  class="MuiGrid-root makeStyles-contentGrid-4 MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
>
  <div
    class="MuiGrid-root makeStyles-subtitle-2 MuiGrid-item"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <h2
        class="MuiTypography-root MuiTypography-h5"
      >
        welcome.subtitle
      </h2>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-item"
  >
    Onboarding
  </div>
  <div
    class="MuiGrid-root MuiGrid-item"
  >
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      href="/en/settings"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        welcome.cta
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
  </div>
  <div
    class="MuiGrid-root makeStyles-boilerplate-5 MuiGrid-item"
  >
    Boilerplate
  </div>
</div>
`;
